File: warehouse/erp_the20/repo/profile.repository.js
// repo/profile.repository.js
// EmployeeProfile: get / getOrCreate / save / list with search & pagination

const { Op } = require("sequelize");

module.exports = function createProfileRepo({ EmployeeProfile }) {
  async function get(user_id) {
    return EmployeeProfile.findOne({ where: { user_id } });
  }

  async function getOrCreate(user_id) {
    const [obj] = await EmployeeProfile.findOrCreate({ where: { user_id } });
    return obj;
  }

  async function saveProfile(user_id, fields) {
    const obj = await getOrCreate(user_id);
    Object.assign(obj, fields);
    await obj.save();
    return obj;
  }

  async function listProfiles({ q = null, limit = 50, offset = 0 } = {}) {
    const where = {};
    if (q) {
      const like = { [Op.iLike]: `%${q}%` };
      Object.assign(where, {
        [Op.or]: [
          { full_name: like },
          { email: like },
          { phone: like },
          { cccd: like },
        ],
      });
    }
    const { rows, count } = await EmployeeProfile.findAndCountAll({
      where,
      order: [["updated_at", "DESC"]],
      limit,
      offset,
    });
    return { total: count, items: rows };
  }

  return { get, getOrCreate, saveProfile, listProfiles };
};

File: warehouse/erp_the20/repositories/profile_repository.py
# erp_the20/repositories/profile_repository.py
from typing import Optional, Tuple, List
from django.db.models import Q
from erp_the20.models import EmployeeProfile

def get(user_id: int) -> Optional[EmployeeProfile]:
    return EmployeeProfile.objects.filter(user_id=user_id).first()

def get_or_create(user_id: int) -> EmployeeProfile:
    obj, _ = EmployeeProfile.objects.get_or_create(user_id=user_id)
    return obj

def save_profile(user_id: int, **fields) -> EmployeeProfile:
    obj = get_or_create(user_id)
    for k, v in fields.items():
        setattr(obj, k, v)
    obj.save()
    return obj

def list_profiles(q: Optional[str] = None, limit: int = 50, offset: int = 0) -> Tuple[int, List[EmployeeProfile]]:
    qs = EmployeeProfile.objects.all().order_by("-updated_at")
    if q:
        qs = qs.filter(
            Q(full_name__icontains=q) |
            Q(email__icontains=q) |
            Q(phone__icontains=q) |
            Q(cccd__icontains=q)
        )
    total = qs.count()
    items = list(qs[offset:offset+limit])
    return total, items
